server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java

```diff
@@ -0,0 +1,209 @@
+  }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getCompactConfig()
+  {
+    return Response.ok(
+        manager.watch(
+            CoordinatorCompactionConfig.CONFIG_KEY,
+            CoordinatorCompactionConfig.class
+        ).get()
+    ).build();
+  }
+
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response setCompactionTaskLimit(
+      @QueryParam("slotRatio") Double compactionTaskSlotRatio,
+      @QueryParam("maxSlots") Integer maxCompactionTaskSlots,
+      @HeaderParam(AuditManager.X_DRUID_AUTHOR) @DefaultValue("") final String author,
+      @HeaderParam(AuditManager.X_DRUID_COMMENT) @DefaultValue("") final String comment,
+      @Context HttpServletRequest req
+  )
+  {
+    CoordinatorCompactionConfig current = manager.watch(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        CoordinatorCompactionConfig.class
+    ).get();
+
+    final CoordinatorCompactionConfig newCompactionConfig;
+    if (current != null) {
+      newCompactionConfig = CoordinatorCompactionConfig.from(current, compactionTaskSlotRatio, maxCompactionTaskSlots);
+    } else {
+      newCompactionConfig = new CoordinatorCompactionConfig(null, compactionTaskSlotRatio, maxCompactionTaskSlots);
+    }
+
+    if (!manager.set(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+    return Response.ok().build();
+  }
+
+  @POST
+  @Path("/{dataSource}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response addOrUpdateCompactionConfig(
+      final DataSourceCompactionConfig newConfig,
+      @PathParam("dataSource") String dataSource,
+      @HeaderParam(AuditManager.X_DRUID_AUTHOR) @DefaultValue("") final String author,
+      @HeaderParam(AuditManager.X_DRUID_COMMENT) @DefaultValue("") final String comment,
+      @Context HttpServletRequest req
+  )
+  {
+    CoordinatorCompactionConfig current = manager.watch(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        CoordinatorCompactionConfig.class
+    ).get();
+
+    final CoordinatorCompactionConfig newCompactionConfig;
+    if (current != null) {
+      final Map<String, DataSourceCompactionConfig> newConfigs = current
+          .getCompactionConfigs()
+          .stream()
+          .collect(Collectors.toMap(DataSourceCompactionConfig::getDataSource, Function.identity()));
+      newConfigs.put(dataSource, newConfig);
+      newCompactionConfig = CoordinatorCompactionConfig.from(current, ImmutableList.copyOf(newConfigs.values()));
+    } else {
+      newCompactionConfig = CoordinatorCompactionConfig.from(ImmutableList.of(newConfig));
+    }
+
+    if (!manager.set(
+        CoordinatorCompactionConfig.CONFIG_KEY,
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
```

Comment:
minor nit: same, please add more details for error if possible.

Reply:
Added.